### ISSUE TITLE
Create DESCRIBE SCHEMA cql files for the whole schema on snapshot

### ISSUE THREAD

**Issue by Michal-Leszczynski** [scylladb/scylla#18271]
As described in https://github.com/scylladb/scylladb/pull/18115#issuecomment-2037281211, when #18115 gets merged, the last step to make the safe for restore schema usable for Scylla Manger is to make Scylla create DESCRIBE SCHEMA files for the whole keyspace schema when creating a snapshot (POST `/storage_service/snapshots` endpoint). Right now only the table DESCRIBE SCHEMA file is created on snapshot.

Also note that UDT schema files make up a DAG that should be respected when restoring the schema via CQL, so there is also a need for Scylla Manager to know this order.

cc: @Jadw1 @tzach @karol-kokoszka @piodul

**Comment by Michal-Leszczynski** [comment]
I believe that this is a blocker for 6.0 release and for working on SM implementation of restoring schema. In case there is no time to include it in a 6.0 release and there is a tested workaround (especially in terms of restoring the DAG of UDTs), please let me know.

**Comment by piodul** [comment]
As far as I know, cross-keyspace references are not allowed, i.e. you can't use a type from one keyspace in a table of another keyspace. I don't think there is a need to toposort them.

Type definitions will have to be topologically sorted as one type can has another type as a field. I'm not sure if this is the case right now, will have to check.

> I believe that this is a blocker for 6.0 release and for working on SM implementation of restoring schema. In case there is no time to include it in a 6.0 release and there is a tested workaround (especially in terms of restoring the DAG of UDTs), please let me know.

We have server-side DESC KEYSPACES/TYPES/FUNCTIONS/AGGREGATES/TABLES WITH INTERNALS which you can just send via CQL to get the statements needed to recreate the items. Of course, the schema can be changed between sending this query and taking the snapshot, so I don't know whether it is a good workaround for you (and maybe we need to implement the toposort anyway).

**Comment by Michal-Leszczynski** [comment]
> We have server-side DESC KEYSPACES/TYPES/FUNCTIONS/AGGREGATES/TABLES WITH INTERNALS which you can just send via CQL to get the statements needed to recreate the items. Of course, the schema can be changed between sending this query and taking the snapshot, so I don't know whether it is a good workaround for you (and maybe we need to implement the toposort anyway).

Ok, I was thinking that server-side DESCRIBE is not supported directly in CQL (so it is only used with cqlsh or in files generated by Scylla). In terms of changing the schema, I guess SM could do the following:
- await schema agreement and record schema version
- make snapshots
- query DESCRIBE SCHEMA
- validate schema agreement and schema version

In case of failure, backup should be retired in some time.

Having that in mind, the only thing that might be missing is the toposort for UDTs in DESCRIBE SCHEMA, but this is not a big deal in terms of 6.0 release.

**Comment by Jadw1** [comment]
> Ok, I was thinking that server-side DESCRIBE is not supported directly in CQL (so it is only used with cqlsh or in files generated by Scylla).

Server-side describe is a regular CQL statement.
To use it in cqlsh, "version" must be >= 4 https://github.com/scylladb/scylla-cqlsh/blob/master/bin/cqlsh.py#L1593 and this is not true for Scylla https://github.com/scylladb/scylladb/blob/master/version.hh#L27.
But in general, you can send any `DESC ...` statement and it will work

**Comment by piodul** [comment]
> Server-side describe is a regular CQL statement. To use it in cqlsh, "version" must be >= 4 https://github.com/scylladb/scylla-cqlsh/blob/master/bin/cqlsh.py#L1593 and this is not true for Scylla https://github.com/scylladb/scylladb/blob/master/version.hh#L27. But in general, you can send any `DESC ...` statement and it will work

Yes, this is a missing feature/bug in our fork of cqlsh: https://github.com/scylladb/scylla-cqlsh/issues/17

**Comment by karol-kokoszka** [comment]
> So this is question about how SM should interact with Scylla 6.0.
Creating all schema files on keyspace snapshot is additional task for Scylla 6.0 release, but it would preserve the current SM approach that always backs up the schema without the need for cluster credentials.
Making SM rely on CQL credentials to back-up schema might result in irrecoverable backups if user forgets about this change in behavior. Or perhaps SM should not allow backups without CQL credentials in such case?

This is very relevant for enterprise clusters.
Scyla Cloud always adds credentials to the manager.
IMHO the best would be to preserve the current behavior and to back up the schema regardless of missing credentials.
We still backup `system-schema`, but we know that it won't be possible to restore it with RAFT.

**Comment by kostja** [comment]
Erm, @karol-kokoszka @Michal-Leszczynski but have you thought about restore? You can't restore CQL just by dropping the files into the data dir. You need to execute it using the same connection/credentials procedure. So what are you actually saving?

**Comment by karol-kokoszka** [comment]
But creating the backup is a different process.
First you create the backup and then you do the restore.
Nothing prevents users from updating the cluster with credentials just before the restore.

If the backup does not keep the schema then you wont bring the schema back doing the restore even if you just updated the credentials in manager.

**Comment by Michal-Leszczynski** [comment]
>  Nothing prevents users from updating the cluster with credentials just before the restore.

It's safer to provide credentials just for the time of restore procedure - especially that restore requires permission to ALTER user table. SM also uses CQL credentials for:
- awaiting schema agreement before backup (right now it's skipped when no credentials are provided)
- distinguishing views from base table (can be skipped and could be fixed by #17958)

**Comment by piodul** [comment]
How about creating a new user with default permissions, and using those credentials? As far as I can tell, such a user only has read-only access to the system tables. It can also call `describe schema with internals`. If you are doing backup then you have read-only access to all data in the cluster, so it's not a privilege escalation.

cc: @nuivall

**Comment by tzach** [comment]
AFAIU the suggestion, we can add such a restore user **only** when doing a restore.
Is it the case? if so, can it be a create cluster parameter?
